### PR TITLE
archive: Refactor `archive_storage` method into subscription

### PR DESCRIPTION
**Pull request by lexnv**
This PR adapts the `archive_storage` implementation from a method to a subscription.

This keeps the archive APIs uniform and consistent.

Builds on: https://github.com/paritytech/polkadot-sdk/pull/5997

cc @paritytech/subxt-team

### CODE REVIEW

File: substrate/client/rpc-spec-v2/src/archive/archive.rs [hunk 1]
```diff
@@ -372,3 +336,24 @@ async fn process_events(rx: &mut mpsc::Receiver<ArchiveStorageDiffEvent>, sink:
 		}
 	}
 }
+
+/// Sends all the events to the sink.
+async fn process_storage_events(rx: &mut mpsc::Receiver<QueryResult>, sink: &mut Subscription) {
+	while let Some(event) = rx.recv().await {
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/archive/archive.rs:342]: perhaps do a select loop here instead to terminate this when the subscription is closed....


```rust
    loop {
        tokio::select! {
        _ = sink.closed() => return,
        event = rx.recv() => { // code to read the event here }
        }
    }
```

File: substrate/client/rpc-spec-v2/src/common/events.rs [hunk 2]
```diff
@@ -105,23 +105,41 @@ pub struct StorageResultErr {
...
+#[serde(tag = "event")]
+pub enum ArchiveStorageEvent {
 	/// Query generated a result.
-	Ok(ArchiveStorageMethodOk),
+	StorageResult(StorageResult),
 	/// Query encountered an error.
-	Err(ArchiveStorageMethodErr),
+	StorageErr(ArchiveStorageMethodErr),
+	/// Operation storage is done.
+	StorageDone,
```
Comment (jsdw) [inline substrate/client/rpc-spec-v2/src/common/events.rs:116]: Shouldn't this match the spec (https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_storage.html#storagedone) ie in the spec we have 3 events, "storage" (getting value back), "storageDone" (finished) and "storageError" for errors?

File: substrate/client/rpc-spec-v2/src/archive/archive.rs [hunk 3]
```diff
@@ -337,24 +299,74 @@ where
...
 }
 
-/// Sends all the events to the sink.
-async fn process_events(rx: &mut mpsc::Receiver<ArchiveStorageDiffEvent>, sink: &mut Subscription) {
-	while let Some(event) = rx.recv().await {
-		if event.is_done() {
-			log::debug!(target: LOG_TARGET, "Finished processing partial trie query");
-		} else if event.is_err() {
-			log::debug!(target: LOG_TARGET, "Error encountered while processing partial trie query");
+/// Sends all the events of the storage_diff method to the sink.
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/archive/archive.rs:311]: A little annoying to have two different functions for `ArchiveStorageDiffEvent` and for `ArchiveStorageEvent` but I have no better suggestion.

It's fine

File: substrate/client/rpc-spec-v2/src/common/events.rs [hunk 4]
```diff
@@ -105,23 +105,41 @@ pub struct StorageResultErr {
...
 /// The result of a storage call.
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-#[serde(untagged)]
-pub enum ArchiveStorageResult {
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "event")]
+pub enum ArchiveStorageEvent {
 	/// Query generated a result.
-	Ok(ArchiveStorageMethodOk),
+	StorageResult(StorageResult),
```
Comment (jsdw) [inline substrate/client/rpc-spec-v2/src/common/events.rs:112]: I also note that `StorageResult` has no way to return `childTrieKey` unless I'm mistaken; should it be possible for this field to be returned?

**Review by niklasad1 — APPROVED** [review]
Looks good, nice to get rid of the static limits :)

File: substrate/client/rpc-spec-v2/src/common/events.rs [hunk 5]
```diff
@@ -105,23 +109,41 @@ pub struct StorageResultErr {
...
-pub enum ArchiveStorageResult {
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "event")]
+pub enum ArchiveStorageEvent {
 	/// Query generated a result.
-	Ok(ArchiveStorageMethodOk),
+	Storage(StorageResult),
 	/// Query encountered an error.
-	Err(ArchiveStorageMethodErr),
+	StorageErr(ArchiveStorageMethodErr),
```
Comment (jsdw) [inline substrate/client/rpc-spec-v2/src/common/events.rs:118]: I think this enum should be called `StorageError` to align with the spec (which uses the events "storage", "storageError" and "storageDone"

**Pull request merged**